README.md
# Intro

React `useBreakpoint` hook to have different values for a variable
based on a breakpoints.

# Demo
https://wintercounter.github.io/use-breakpoint

# Install

```bash
npm i @w11r/use-breakpoint
```

# Setup

## Add provider

Add `BreakpointProvider` in your React tree.

```jsx
import { BreakpointProvider } from '@w11r/use-breakpoint'

...
    <BreakpointProvider>
        ...
    </BreakpointProvider>
...
```

## Default breakpoints

The following default breakpoints are being used, including several shorthands.

```js
const breakpoints = {
    micro: [0, 375],
    mi: [0, 375],
    mobile: [376, 639],
    m: [376, 639],
    tablet: [640, 1023],
    t: [640, 1023],
    small: [1024, 1439],
    s: [1024, 1439],
    medium: [1440, 1919],
    med: [1440, 1919],
    large: [1920, 10000],
    l: [1920, 10000],
    // Multi range
    device: [0, 1023],
    d: [0, 1023],
    smallDevice: [0, 639],
    sd: [0, 639]
}
```

## Override default settings

```js
import { setup, breakpoints } from '@w11r/use-breakpoint'

setup({
    breakpoints: {
        // Extend default values
        ...breakpoints,
        alienDevice: [342, 43534] // from, to
    }
})
```

# Usage

## With passing values

```js
useBreakpoint(defaultValue, breakpointValues)

// breakpointValues: array of breakpoint based values
[
  ['mobile', 300],
  ['tablet', 400]
]

// In case you have a single breakpoint value, `['mobile', 300]` is enough instead of `[['mobile', 300]]`
```

## Without passing values

In case you don't specify any value to the hook, it'll return a generated
object including boolean values for each breakpoint keys that's being
defined in options.

It'll return the following object with the basic setup.

```
{
      isLandscape: false,
      isPortrait: true,
      isHDPI: false,
      isMicro: false,
      isMobile: true,
      isTablet: false,
      isSmall: false,
      isMedium: false,
      isLarge: false,
      'is-Micro': false,
      'is|Micro': false,
      'isMicro+': true,
      'is-Micro+': true,
      'is|Micro+': true,
      'isMicro-': false,
      'is-Micro-': false,
      'is|Micro-': false,
      'is-Mobile': true,
      'is|Mobile': true,
      'isMobile+': true,
      'is-Mobile+': true,
      'is|Mobile+': true,
      'isMobile-': true,
      'is-Mobile-': true,
      'is|Mobile-': true,
      'is-Tablet': false,
      'is|Tablet': false,
      'isTablet+': false,
      'is-Tablet+': false,
      'is|Tablet+': false,
      'isTablet-': true,
      'is-Tablet-': true,
      'is|Tablet-': true,
      'is-Small': false,
      'is|Small': false,
      'isSmall+': false,
      'is-Small+': false,
      'is|Small+': false,
      'isSmall-': true,
      'is-Small-': true,
      'is|Small-': true,
      'is-Medium': false,
      'is|Medium': false,
      'isMedium+': false,
      'is-Medium+': false,
      'is|Medium+': false,
      'isMedium-': true,
      'is-Medium-': true,
      'is|Medium-': true,
      'is-Large': false,
      'is|Large': false,
      'isLarge+': false,
      'is-Large+': false,
      'is|Large+': false,
      'isLarge-': true,
      'is-Large-': true,
      'is|Large-': true
    }
```

Usage
```js
const { isMobile } = useBreakpoint()

// The above is basically a replacement for
const isMobile = useBreakpoint(false, ['mobile', true])
```

> You can also access the values with suffix and prefix, but you need
> to rename the variables because it contains invalid character:
> `const { 'isMobile+': isMobile } = useBreakpoint()`

Component example

```jsx
import useBreakpoint from '@w11r/use-breakpoint'

const MyCmp = () => {
    const columns = useBreakpoint([1,2], ['mobile', [2,1]])

    return <Grid cols={columns} />
}

// Or using inline
const MyCmp = () => {
    return <Grid cols={useBreakpoint([1,2], ['mobile', [2,1]])} />
}
```

> _Rules-of-Hooks_ are still true in this case as well. Make sure
> your component will __ALWAYS__ run it without any condition!

## Modifiers

All breakpoint names coming with modifiers included.

### Orientation prefix

- `` (none): all
- `-`: Landscape
- `|`: Portrait

### Range suffix

You can also control your value to behave as `and up` and `and down`.

- `` (none): all
- `+`: `and up`
- `-`: `and down`

### Examples

- `['|mobile', 300]`: on mobile, on portrait
- `['|mobile+', 300]`: on mobile and up, on portrait
- `['mobile+', 300]`: on mobile and up, both portrait and landscape
- `['mobile', 300]`: on mobile, both portrait and landscape
- `['tablet-', 300]`: on tablet and below, both portrait and landscape
- `['mobile-', 300]`: on mobile and down, both portrait and landscape

## Generate Media Query strings using the same logic

It is useful when you just need simple CSS. For example with `styled-components`:

```js
import styled from 'styled-components'
import { mediaQuery } from '@w11r/use-breakpoint'

const mediaQueryString = mediaQuery(
    ['mobile-', `width: 100%;`]
)

const Box = styled.div`
    ${mediaQuery(
        ['mobile-', `width: 100%;`]
    )}
`

// You can still use multiple queries at once just like with the hook:
mediaQuery([['mobile-', `width: 100%`], ['medium+', `width: 50%`]])
```

This utility comes with additional support for `shorthands` which is configurable through `options`.

```js
setup({
    shorthands: {
        foo: '@media (wow: bar) and (hmm: khm)'
    }
})

mediaQuery(['foo', value])
```

Additionally, it supports `dark` and `light` mode prefixes:
- `(`: Light mode
- `)`: Dark mode

```js
mediaQuery([')mobile', 'content'])
// @media screen and (min-width: 376px) and (max-width: 639px) and (prefers-color-scheme: dark) { content }

mediaQuery([')', 'content'])
// @media screen and (prefers-color-scheme: dark) { content }
```

# FAQ

## Is there any best practice suggestion?

Yes! Use as fewer hooks as possible. It's always faster to have a single
`isMobile` variable and have simple conditions based on it. It's even better
if you can solve your size related cases using pure CSS Media Queries.

## Why not using an `Object`? Why the `Array` structure?

Object's cannot guarantee the order of the defined keys. It is crucial
to check for values in the correct order because `useBreakpoint` uses
eager evaluation and `mediaQuery` must maintain the defined order of
the generated Media Queries.

## Which rule is being prioritized?

The hook uses _eager_ evaluation, so the first truthy breakpoint value
gets returned.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [2.0.0](https://github.com/wintercounter/use-breakpoint/compare/v1.7.1...v2.0.0) (2021-04-21)


### ⚠ BREAKING CHANGES

* **context:** Add provider to your React tree

```js
import { BreakpointProvider } from '@w11r/use-breakpoint'
```

### Features

* **context:** move resize logic into context provider ([74e3531](https://github.com/wintercounter/use-breakpoint/commit/74e3531ca152e63d1af0632d769d3bd2c8e5f796))
* **mediaQuery:** add support for prefix only breakpoints ([436e6e7](https://github.com/wintercounter/use-breakpoint/commit/436e6e7cd50b5a8575170e3209e98a4751af4541))
* **mediaQuery:** add support for shorthands ([2a58497](https://github.com/wintercounter/use-breakpoint/commit/2a58497dd64130a4eb81f86a6cce8dcc2fe7f7a4))

## [1.8.0](https://github.com/wintercounter/use-breakpoint/compare/v1.7.1...v1.8.0) (2021-01-02)


### Features

* **mediaQuery:** add support for prefix only breakpoints ([436e6e7](https://github.com/wintercounter/use-breakpoint/commit/436e6e7cd50b5a8575170e3209e98a4751af4541))
* **mediaQuery:** add support for shorthands ([2a58497](https://github.com/wintercounter/use-breakpoint/commit/2a58497dd64130a4eb81f86a6cce8dcc2fe7f7a4))

### [1.7.1](https://github.com/wintercounter/use-breakpoint/compare/v1.7.0...v1.7.1) (2020-12-02)


### Bug Fixes

* **useBreakpoints:** adding extra check on window so nextjs won't complain ([e35efdb](https://github.com/wintercounter/use-breakpoint/commit/e35efdbcdf76878445a92934327c21e36fc1c187))
* **useBreakpoints:** NextJS support ([9150685](https://github.com/wintercounter/use-breakpoint/commit/915068574328c50bf77aec9c0746021f98fac3d6))

## [1.7.0](https://github.com/wintercounter/use-breakpoint/compare/v1.6.0...v1.7.0) (2020-07-01)


### Features

* add support for media expressions directly from propless useBreakpoint ([b188a49](https://github.com/wintercounter/use-breakpoint/commit/b188a49f65e3f15a03e8ccabd57f5d9f1988ad87))


### Bug Fixes

* use correct cache key ([84e4a5c](https://github.com/wintercounter/use-breakpoint/commit/84e4a5c41765f2a22b80e1a98381a97c79e59678))

## [1.6.0](https://github.com/wintercounter/use-breakpoint/compare/v1.5.0...v1.6.0) (2020-06-21)


### Features

* add multi-range breakpoints ([29c2b9f](https://github.com/wintercounter/use-breakpoint/commit/29c2b9f03899c756c7a6103bb28703112a300e36))

## [1.5.0](https://github.com/wintercounter/use-breakpoint/compare/v1.4.3...v1.5.0) (2020-06-06)


### Features

* add cjs version ([15e436c](https://github.com/wintercounter/use-breakpoint/commit/15e436c543958cb339980ea257462c31452beb91))

### [1.4.3](https://github.com/wintercounter/use-breakpoint/compare/v1.4.2...v1.4.3) (2020-05-15)


### Bug Fixes

* missing dependencies ([cbf6f3a](https://github.com/wintercounter/use-breakpoint/commit/cbf6f3a63487b134218f8a374d1d982c7a8d08f0))

### [1.4.2](https://github.com/wintercounter/use-breakpoint/compare/v1.4.1...v1.4.2) (2020-05-15)


### Bug Fixes

* stop relying on orientation ([8e5b24a](https://github.com/wintercounter/use-breakpoint/commit/8e5b24a61f90f5a91f7976bbbac33045454bd853))

### [1.4.1](https://github.com/wintercounter/use-breakpoint/compare/v1.4.0...v1.4.1) (2020-03-25)


### Bug Fixes

* **useBreakpoint:** match logic to Media Queries logic ([b7b3d56](https://github.com/wintercounter/use-breakpoint/commit/b7b3d56ae55859d97005d34a08d70075852789cd))
* useBreakpoint input types ([813c8e9](https://github.com/wintercounter/use-breakpoint/commit/813c8e9cb57ab7f13b69b2c4f41552cc02e30be4))

## [1.4.0](https://github.com/wintercounter/use-breakpoint/compare/v1.3.3...v1.4.0) (2020-03-05)


### Features

* **mediaQuery:** add support for Object type output ([4025772](https://github.com/wintercounter/use-breakpoint/commit/402577293a8818033cd3e8583892da03498e1bf8))
* **options:** add support for short names ([0a4d398](https://github.com/wintercounter/use-breakpoint/commit/0a4d3984f4cf7eb435b69e29dd345b27661b7e44))

### [1.3.3](https://github.com/wintercounter/use-breakpoint/compare/v1.3.2...v1.3.3) (2020-02-29)


### Bug Fixes

* **ssr:** temp solution to allow ssr till v2 ([4e1e5a9](https://github.com/wintercounter/use-breakpoint/commit/4e1e5a9b612f38bf1a5410cf30c23b2401aae275))

### [1.3.2](https://github.com/wintercounter/use-breakpoint/compare/v1.2.1...v1.3.2) (2020-02-29)


### Bug Fixes

* **default:** handle orientation correctly ([b1f2ff4](https://github.com/wintercounter/use-breakpoint/commit/b1f2ff4c1fe7a4e98fce75d3ca406529c59fa53f))
* **ssr:** temp solution to allow ssr till v2 ([4c62d7c](https://github.com/wintercounter/use-breakpoint/commit/4c62d7cf3f638c1de84c989d8561216176d940ec))

### [1.3.1](https://github.com/wintercounter/use-breakpoint/compare/v1.2.1...v1.3.1) (2020-02-07)

### Bug Fixes

* **mediaQuery:** emit all size declarations not only the first one ([ac7ecfd](https://github.com/wintercounter/use-breakpoint/commit/ac7ecfdc27336bb3738c0fcc967efb76b798ddd6))

## [1.3.0](https://github.com/wintercounter/use-breakpoint/compare/v1.2.0...v1.3.0) (2020-01-16)


### Features

* add support for generating Media Query strings ([b4cac9f](https://github.com/wintercounter/use-breakpoint/commit/b4cac9fea59f7e83e17683b839229ee9dfbe164e))


### Bug Fixes

* TS declaration for the case when no values are provided ([ad23da4](https://github.com/wintercounter/use-breakpoint/commit/ad23da4d7e8c5acf8672661e0ba50aae497fc260))
* **default:** handle orientation correctly ([b1f2ff4](https://github.com/wintercounter/use-breakpoint/commit/b1f2ff4c1fe7a4e98fce75d3ca406529c59fa53f))

## [1.2.0](https://github.com/wintercounter/use-breakpoint/compare/v1.1.1...v1.2.0) (2020-01-06)


### Features

* add support for generating Media Query strings ([b4cac9f](https://github.com/wintercounter/use-breakpoint/commit/b4cac9fea59f7e83e17683b839229ee9dfbe164e))


### Bug Fixes


src/index.ts
import useBreakpoint from './useBreakpoint'
import useResize from './useResize'
import mediaQuery from './mediaQuery'

export * from './provider'

export const UP = '+'
export const DOWN = '-'
export const LANDSCAPE = '-'
export const PORTRAIT = '|'
export const LIGHT = '('
export const DARK = ')'

interface IOptions {
    breakpoints: {
        [key: string]: number[]
    }
    shorthands?: {
        [key: string]: string
    }
}

export let options

export const setup = function (opts: IOptions) {
    Object.entries(opts.breakpoints).forEach(([name, [from, to]]) =>
        [
            ['', [from, to]],
            [UP, [from, 10000]],
            [DOWN, [0, to]]
        ].forEach(([symbol, fromTo]) =>
            ['', LANDSCAPE, PORTRAIT, LIGHT, DARK].forEach(prefix => {
                // eslint-disable-next-line
                opts.breakpoints[`${prefix}${name}${symbol}`] = fromTo as [number, number]
            })
        )
    )
    // Only-prefix support
    Object.assign(opts.breakpoints, {
        [LANDSCAPE]: [true],
        [PORTRAIT]: [true],
        [LIGHT]: [true],
        [DARK]: [true]
    })

    options = opts
}

export const breakpoints = {
    micro: [0, 375],
    mi: [0, 375],
    mobile: [376, 639],
    m: [376, 639],
    tablet: [640, 1023],
    t: [640, 1023],
    small: [1024, 1439],
    s: [1024, 1439],
    medium: [1440, 1919],
    med: [1440, 1919],
    large: [1920, 10000],
    l: [1920, 10000],
    // Multi range
    device: [0, 1023],
    d: [0, 1023],
    smallDevice: [0, 639],
    sd: [0, 639]
}
const shorthands = {
    dark: '@media screen and (prefers-color-scheme: dark)',
    light: '@media screen and (prefers-color-scheme: light)',
    hdpi: '@media screen and (min-resolution: 192dpi)'
}

setup({ breakpoints, shorthands })

export default useBreakpoint
export { useResize, mediaQuery }
